pkg/member/member_add.go

```diff
@@ -0,0 +1,158 @@
+		cli, _ := clientFactory.NewCluster()
+		_, err := cli.MemberAddAsLearner(memAddCtx, []string{memberURL})
+		cancel()
+		cli.Close()
+
+		if err != nil {
+			logger.Warn("Error adding member as a learner: ", err)
+		}
+		if err == nil || strings.Contains(rpctypes.ErrGRPCPeerURLExist.Error(), err.Error()) {
+			logger.Info("Added member to cluster as a learner")
+			break //TODO: why not just return here?
+		}
+		if strings.Contains(rpctypes.ErrGRPCPeerURLExist.Error(), err.Error()) {
+			logger.Info("Member already part of etcd cluster")
+			break
+		}
+
+		logger.Info("Could not as member as learner due to: ", err)
+		logger.Info("Trying again in 5 seconds... ")
+		timer := time.NewTimer(RetryPeriod)
+		<-timer.C
+		timer.Stop()
+	}
+
+	return nil
+}
+
+// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+func IsMemberInCluster(logger *logrus.Logger) bool {
+	//Create etcd client
+	// TODO: Use secure transport
+	clientFactory := etcdutil.NewFactory(brtypes.EtcdConnectionConfig{
+		Endpoints:         []string{os.Getenv("ETCD_ENDPOINT")},
+		InsecureTransport: true, //TODO: is it right to use insecure transport?
+	})
+
+	// TODO: should use a retry mechanism here
+	cli, _ := clientFactory.NewCluster()
+	defer cli.Close()
+	logger.Info("Etcd client created")
+
+	// List members in cluster
+	memListCtx, cancel := context.WithTimeout(context.TODO(), brtypes.DefaultEtcdConnectionTimeout)
+	etcdMemberList, err := cli.MemberList(memListCtx)
+	defer cancel()
+	if err != nil {
+		logger.Warn("Could not list any etcd members", err)
+		return true
+	}
+
+	for _, y := range etcdMemberList.Members {
+		if y.Name == os.Getenv("POD_NAME") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getMemberURL() string {
+	end := strings.Split(os.Getenv("ETCD_ENDPOINT"), "//")
+	memberURL := end[0] + "//" + os.Getenv("POD_NAME") + "." + end[1]
+	return memberURL
+}
+
+// PromoteMember promotes an etcd member from a learner to a voting member of the cluster. This will succeed only if its logs are caught up with the leader
+func PromoteMember(ctx context.Context, logger *logrus.Entry) {
+	for {
+		// TODO: Use secure transport
+		clientFactory := etcdutil.NewFactory(brtypes.EtcdConnectionConfig{
+			Endpoints:         []string{os.Getenv("ETCD_ENDPOINT")},
+			InsecureTransport: true,
+		})
+		cli, _ := clientFactory.NewCluster()
+
+		//List all members in the etcd cluster
+		//Member URL will appear in the memberlist call response as soon as the member has been added to the cluster as a learner
+		//However, the name of the member will appear only if the member has started running
+		memListCtx, memListCtxcancel := context.WithTimeout(context.TODO(), brtypes.DefaultEtcdConnectionTimeout)
+		etcdList, memListErr := cli.MemberList(memListCtx)
+		memListCtxcancel()
+
+		if memListErr != nil {
+			logger.Info("error listing members: ", memListErr)
+			cli.Close()
+			continue
```

Comment:
is it possible to have unrecoverable errors, if yes then you will end up in an infinite loop. I would highly recommend not to do retries in this file. Return proper errors and then let the caller handle the retry.
If you really wish to retry in this file, then have 2 functions one which does not retry and one which retries and delegates to the function which does not retry.